pennylane_lightning/core/src/simulators/lightning_qubit/measurements/MeasurementKernels.hpp

```diff
@@ -0,0 +1,350 @@
+#define PROBS_CORE_SUM_7_8(x3, x4, x5, x6)                                     \
+    PROBS_CORE_SUM_7_4(0, x3, x4, x5, x6) PROBS_CORE_SUM_7_4(1, x3, x4, x5, x6)
+#define PROBS_CORE_SUM_7_16(x4, x5, x6)                                        \
+    PROBS_CORE_SUM_7_8(0, x4, x5, x6) PROBS_CORE_SUM_7_8(1, x4, x5, x6)
+#define PROBS_CORE_SUM_7_32(x5, x6)                                            \
+    PROBS_CORE_SUM_7_16(0, x5, x6) PROBS_CORE_SUM_7_16(1, x5, x6)
+#define PROBS_CORE_SUM_7_64(x6)                                                \
+    PROBS_CORE_SUM_7_32(0, x6) PROBS_CORE_SUM_7_32(1, x6)
+#define PROBS_CORE_SUM_7                                                       \
+    if constexpr (n_wires == 7) {                                              \
+        i0 = PROBS_CORE_DEF_I0000000;                                          \
+        PROBS_CORE_SUM_7_64(0);                                                \
+        PROBS_CORE_SUM_7_64(1);                                                \
+    }
+#define PROBS_CORE_SUM_8_2(x1, x2, x3, x4, x5, x6, x7)                         \
+    PROBS_CORE_DEF_Ixy8(0, x1, x2, x3, x4, x5, x6, x7)                         \
+        PROBS_CORE_DEF_Ixy8(1, x1, x2, x3, x4, x5, x6, x7)
+#define PROBS_CORE_SUM_8_4(x2, x3, x4, x5, x6, x7)                             \
+    PROBS_CORE_SUM_8_2(0, x2, x3, x4, x5, x6, x7)                              \
+    PROBS_CORE_SUM_8_2(1, x2, x3, x4, x5, x6, x7)
+#define PROBS_CORE_SUM_8_8(x3, x4, x5, x6, x7)                                 \
+    PROBS_CORE_SUM_8_4(0, x3, x4, x5, x6, x7)                                  \
+    PROBS_CORE_SUM_8_4(1, x3, x4, x5, x6, x7)
+#define PROBS_CORE_SUM_8_16(x4, x5, x6, x7)                                    \
+    PROBS_CORE_SUM_8_8(0, x4, x5, x6, x7) PROBS_CORE_SUM_8_8(1, x4, x5, x6, x7)
+#define PROBS_CORE_SUM_8_32(x5, x6, x7)                                        \
+    PROBS_CORE_SUM_8_16(0, x5, x6, x7) PROBS_CORE_SUM_8_16(1, x5, x6, x7)
+#define PROBS_CORE_SUM_8_64(x6, x7)                                            \
+    PROBS_CORE_SUM_8_32(0, x6, x7) PROBS_CORE_SUM_8_32(1, x6, x7)
+#define PROBS_CORE_SUM_8_128(x6)                                               \
+    PROBS_CORE_SUM_8_64(0, x6) PROBS_CORE_SUM_8_64(1, x6)
+#define PROBS_CORE_SUM_8                                                       \
+    if constexpr (n_wires == 8) {                                              \
+        i0 = PROBS_CORE_DEF_I00000000;                                         \
+        PROBS_CORE_SUM_8_128(0);                                               \
+        PROBS_CORE_SUM_8_128(1);                                               \
+    }
+
+namespace Pennylane::LightningQubit::Measures {
+
+template <class PrecisionT>
+auto probs_bitshift_generic(const std::complex<PrecisionT> *arr,
+                            const std::size_t num_qubits,
+                            const std::vector<std::size_t> &wires) {
```

Comment:
The generic implementation isn't used for now because it is less performant, but it may be more performant when combined with diagonalization gates.